sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,890 @@
+      spillBatches, spillCallback)
+
+    // Update the remaining partitions
+    poss.foreach(remainingPartIdsAndSizes.remove)
+    (ids, retBatch)
+  }
+
+  private[this] def nextPartitionIdsAndPoss(): Array[(Int, Int)] = {
+    val ret = ArrayBuffer.empty[(Int, Int)]
+    var totalBatches = 0
+    // always append the first one
+    val (firstId, firstBatchCnt) = remainingPartIdsAndSizes.head
+    ret += ((firstId, 0))
+    totalBatches += firstBatchCnt
+    // For each output, try to collect small nonempty partitions to reach "avgNumBatches"
+    // batches as much as possible.
+    // e.g. We have 5 partitions as below. The first partition has two batches, and the
+    // other two partitions have one batch for each.
+    //    p1: (3,4)
+    //    p2: (1,2), (11, 22)
+    //    p3: (5,6)
+    //    p4: <empty>
+    //    p5: <empty>
+    // Then the p1 and p3 will be put together in one output.
+    //    next: (3,4), (5,6)
+    //    next: (1,2), (11, 22)
+    //    next: <empty>
+    //    next: <empty>
+    if (firstBatchCnt > 0) {
+      remainingPartIdsAndSizes.tail.zipWithIndex.foreach { case ((id, batchCnt), pos) =>
+        // Do not coalesce empty partitions, and output them one by one.
+        if (batchCnt > 0 && (totalBatches + batchCnt) <= avgNumBatches) {
+          ret += ((id, pos))
+          totalBatches += batchCnt
+        }
+      }
+    }
+    ret.toArray
+  }
+}
+
+/**
+ * An utils class that will take over the two input batches.
+ */
+class BatchPair(
+    private var build: Option[ColumnarBatch],
+    private var stream: Option[ColumnarBatch]) extends AutoCloseable {
+
+  /**
+   * Whether the batch pair is empty.
+   * A pair is empty when both left and right batches are empty.
+   */
+  def isEmpty: Boolean = build.isEmpty && stream.isEmpty
+
+  /**
+   * Get the two batches as a Tuple(build, stream).
+   */
+  def get: (Option[ColumnarBatch], Option[ColumnarBatch]) = {
+    (build, stream)
+  }
+
+  /**
+   * Release the two batches as a Tuple(build, stream).
+   */
+  def release: (Option[ColumnarBatch], Option[ColumnarBatch]) = {
+    val ret = (build, stream)
+    build = None
+    stream = None
+    ret
+  }
+
+  override def close(): Unit = {
+    build.foreach(_.safeClose(new Exception))
+    stream.foreach(_.close())
+  }
+}
+
+/**
+ * Iterator that returns a pair of batches (build batch, stream batch) generated by
+ * by sub-partitioning algorithm when each call to "next".
+ * It will skip the empty pairs by default. Set "skipEmptyPairs" to false to also get
+ * the empty pairs.
+ */
+class GpuBatchPairSubPartitionIterator(
+    buildIter: Iterator[ColumnarBatch],
+    boundBuildKeys: Seq[GpuExpression],
+    streamIter: Iterator[ColumnarBatch],
+    boundStreamKeys: Seq[GpuExpression],
+    numPartitions: Int,
+    spillCallback: SpillCallback,
+    skipEmptyPairs: Boolean = true)
+  extends Iterator[BatchPair] with Arm with AutoCloseable {
```

Comment:
I personally think that a much cleaner design would be to split things at a different level.

We start off with buildIter and streamIter.  We check to see if the size of buildIter is too large (similar to what this patch is doing).  If it is, then we repartition both sides.  Very similar to what the code just below this is doing.  But instead of trying to combine them together into BatchPair's that are joined.  We go back to the original join Iterator code.

We figure out what partitions we want to combine using the build side and create one iterator for that group of partitions.  We then create a corresponding Iterator for the same group of partitions on the stream side and run them through the original join iterator to produce an output iterator. Then when that iterator says it would be done, we go on to create Iterators for the next group of partitions.

I am happy to do an example if you want me to.

Reply:
I guess I know what you suggest here. 

Each comibnation of the partitioned-and-coalesced build batch and its corresponding stream batches is equal to one original join iterator, meaning inside the sub join iterator, it can leverage the original join iterator to do joins.

Reply:
Yes.  If that is going to be a lot of work I am fine with keeping what you have today. I just think it is going to help with any possible solution to a full outer join and also make it much simpler to try and put in other optimizations in the future while maintaining the core of the join code as the same.

Reply:
Nice suggestion, I will update it accordingly.

Reply:
Updated